### ISSUE TITLE
Bugsnag terminating console apps

### ISSUE THREAD

**Issue by mrmoses** [bugsnag/bugsnag-dotnet#112]
Including Bugsnag in a console application will sometimes terminate the entire application.

We have a few console applications that run via Scheduled Tasks. We have them log to a file, and added logging to Bugsnag as well. After adding Bugsnag, sometimes (1 out of ~30 times) the applications do not finish. The log file just stops in the middle, no errors. No error in Bugsnag either.

Found this in Event Viewer:

> Application: MyApp.exe
> Framework Version: v4.0.30319
> Description: The process was terminated due to an unhandled exception.
> Exception Info: System.NullReferenceException
>    at Bugsnag.WebRequest.EndSend(System.IAsyncResult)
>    at Bugsnag.ThreadQueueDelivery.ReportCallback(System.IAsyncResult)
>    at Bugsnag.WebRequest.ReadCallback(System.IAsyncResult)
>    at System.Net.LazyAsyncResult.Complete(IntPtr)
>    at System.Net.ContextAwareResult.CompleteCallback(System.Object)
>    at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
>    at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
>    at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
>    at System.Net.ContextAwareResult.Complete(IntPtr)
>    at System.Net.LazyAsyncResult.ProtectedInvokeCallback(System.Object, IntPtr)
>    at System.Net.HttpWebRequest.SetResponse(System.Exception)
>    at System.Net.HttpWebRequest.SetAndOrProcessResponse(System.Object)
>    at System.Net.ConnectionReturnResult.SetResponses(System.Net.ConnectionReturnResult)
>    at System.Net.Connection.ReadComplete(Int32, System.Net.WebExceptionStatus)
>    at System.Net.Connection.ReadCallback(System.IAsyncResult)
>    at System.Net.Connection.ReadCallbackWrapper(System.IAsyncResult)
>    at System.Net.LazyAsyncResult.Complete(IntPtr)
>    at System.Net.LazyAsyncResult.ProtectedInvokeCallback(System.Object, IntPtr)
>    at System.Net.Security._SslStream.ReadHeaderCallback(System.Net.AsyncProtocolRequest)
>    at System.Net.AsyncProtocolRequest.CompleteRequest(Int32)
>    at System.Net.FixedSizeReader.CheckCompletionBeforeNextRead(Int32)
>    at System.Net.FixedSizeReader.ReadCallback(System.IAsyncResult)
>    at System.Net.LazyAsyncResult.Complete(IntPtr)
>    at System.Net.ContextAwareResult.CompleteCallback(System.Object)
>    at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
>    at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
>    at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
>    at System.Net.ContextAwareResult.Complete(IntPtr)
>    at System.Net.LazyAsyncResult.ProtectedInvokeCallback(System.Object, IntPtr)
>    at System.Net.Sockets.BaseOverlappedAsyncResult.CompletionPortCallback(UInt32, UInt32, System.Threading.NativeOverlapped*)
>    at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32, UInt32, System.Threading.NativeOverlapped*)

There is no errors after removing Bugsnag, so I don't think the application is erroring and Bugsnag errors trying to log that error. It seems like Bugsnag is doing something other than trying to log an error, and is failing to do whatever that may be, which kills the entire application.

We do use Breadcrumbs a little bit; does leaving a breadcrumb trigger a WebRequest that might be failing?

**Comment by matdavies** [comment]
I don't think it is only console apps that are affected.  We're seeing the same thing in our web application:

```
Application: w3wp.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.NullReferenceException
   at Bugsnag.WebRequest.EndSend(System.IAsyncResult)
   at Bugsnag.ThreadQueueDelivery.ReportCallback(System.IAsyncResult)
   at System.Net.LazyAsyncResult.Complete(IntPtr)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Net.ContextAwareResult.Complete(IntPtr)
   at System.Net.LazyAsyncResult.ProtectedInvokeCallback(System.Object, IntPtr)
   at System.Net.HttpWebRequest.SetResponse(System.Exception)
   at System.Net.ConnectionReturnResult.SetResponses(System.Net.ConnectionReturnResult)
   at System.Net.Connection.ReadComplete(Int32, System.Net.WebExceptionStatus)
   at System.Net.LazyAsyncResult.Complete(IntPtr)
   at System.Net.LazyAsyncResult.ProtectedInvokeCallback(System.Object, IntPtr)
   at System.Net.AsyncProtocolRequest.CompleteWithError(System.Exception)
   at System.Net.FixedSizeReader.ReadCallback(System.IAsyncResult)
   at System.Net.LazyAsyncResult.Complete(IntPtr)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Net.ContextAwareResult.Complete(IntPtr)
   at System.Net.LazyAsyncResult.ProtectedInvokeCallback(System.Object, IntPtr)
   at System.Net.Sockets.BaseOverlappedAsyncResult.CompletionPortCallback(UInt32, UInt32, System.Threading.NativeOverlapped*)
   at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32, UInt32, System.Threading.NativeOverlapped*)

```

In this case it's possible that BugSnag was attempting to report an error (haven't delved further into that), but none the less, brings the whole app crashing down...

**Comment by yousif-bugsnag** [comment]
Closing this as we believe the issue was fixed in [v2.2.0](https://github.com/bugsnag/bugsnag-dotnet/releases/tag/v2.2.0), but will reopen again if we receive any further reports.
